src/diffusers/video_processor.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Union
+
+import numpy as np
+import PIL
+import torch
+
+from .image_processor import VaeImageProcessor
+
+
+class VideoProcessor(VaeImageProcessor):
+    r"""Simple video processor."""
+
+    def tensor2vid(self, video: torch.FloatTensor, output_type: str = "np"):
+        """Converts a video tensor to a list of frames for export."""
+        batch_size = video.shape[0]
+        outputs = []
+        for batch_idx in range(batch_size):
+            batch_vid = video[batch_idx].permute(1, 0, 2, 3)
+            batch_output = self.postprocess(batch_vid, output_type)
+            outputs.append(batch_output)
+
+        if output_type == "np":
+            outputs = np.stack(outputs)
+        elif output_type == "pt":
+            outputs = torch.stack(outputs)
+        elif not output_type == "pil":
+            raise ValueError(f"{output_type} does not exist. Please choose one of ['np', 'pt', 'pil']")
+
+        return outputs
+
+    def preprocess_video(self, video: List[Union[PIL.Image.Image, np.ndarray, torch.Tensor]]) -> torch.FloatTensor:
+        """Preprocesses input video(s)."""
+        supported_formats = (np.ndarray, torch.Tensor, PIL.Image.Image)
```

Comment:
let's first decide what should be video input format we accept here, I think

1. list of list of images 
2. list of images (in that case, we expand it as list of list of images)
3. list of list of tensors (dim4)
4. list of tensors (dim5)
5. list of list of numpy arrays (dim4)
6. list of numpy arrays (dim5)

cc @DN6 here, anything we would add or remove?

Reply:
cc @a-r-r-o-w too since you worked on a lot of video pipelines

Reply:
Okay, so I think all of those are covered. `supported_formats` here refer to the core base format which can be used to create lists, 4d or 5d tensors, etc. 

Let me provide concrete lines of code that I think ensure all the formats listed above are supported: 

1. https://github.com/huggingface/diffusers/blob/7561c1e3a15821c8bb930d13e7392b3a743475d2/src/diffusers/video_processor.py#L58 and

https://github.com/huggingface/diffusers/blob/7561c1e3a15821c8bb930d13e7392b3a743475d2/src/diffusers/video_processor.py#L70

2. and 3. https://github.com/huggingface/diffusers/blob/459843afc856f95c00fd87346f46cd674f99741f/src/diffusers/video_processor.py#L80
4. and 5. https://github.com/huggingface/diffusers/blob/459843afc856f95c00fd87346f46cd674f99741f/src/diffusers/video_processor.py#L64

Reply:
Thanks for looping me in. I think that `list[list[image]]`, `list[fchw_tensor]` and `bfchw_tensor` are most common to end up with as a user, and all cases you mentioned seem to be handled well and the code makes sense to me.

A small explanation of expected inputs to video processor in the docstrings, order of tensor dims, and some more documentation, like in image processor, would be helpful for newer users imo.

Reply:
Thank you, @a-r-r-o-w! Added a bit of documentation and type annotation to hopefully make it clearer. LMK your thoughts.

Reply:
Looks great, thanks

Reply:
a lot of the work here is overlapping with `preprocess` though 

once we make sure the inputs are accepted as one of below (and expanded)
* list of list of images (or a list of images, in that case, we expand it as list of list of images)
* list of list of 4d tensors (or a list of 4d tensor, in this case, we expand)
* list of 5d tensors (or 5d tensor, we expand)
* list of list of 4d numpy arrays (or a list of 4d array, we expand)
* list of 5d numpy arrays (or 5d numpy array, we expand)



can we try:
```python
video = [self.image_processor.preprocess(vid) for vid in videos]
```

after that we check if all the tensors has same shape in the list and throw an error if not

Reply:
> a lot of the work here is overlapping with preprocess though

Help me understand this a bit better? From what I understand, the current `preprocess_video()` is first checking if we have the inputs in the accepted format and is performing the expansion if needed. And then it's passing off the `video` to `preprocess()` like you suggested:

https://github.com/huggingface/diffusers/blob/c5d22e68d1480fddf2725a975778b2c69adfc84d/src/diffusers/video_processor.py#L125

Reply:
The only overlap I see is this:
https://github.com/huggingface/diffusers/blob/c5d22e68d1480fddf2725a975778b2c69adfc84d/src/diffusers/video_processor.py#L106

But I see this more as a safeguard rather than an overlap.

Reply:
i think something like this should work (made up code, but roughly the logic)
```python
if isinstance(video, supported_formats):
    video = [video]
if isinstance(video[0],  PIL.Image.Image)or if isinstance(video[0], ( np.ndarray, torch.tensor) and video[0].ndim ==4):
    video = [video]

video = torch.stack([self.preprocess(f) for f in video],..)
video = video.permute(0, 2, 1, 3, 4)
```

How about let's first add tests like this https://github.com/huggingface/diffusers/blob/main/tests/others/test_image_processor.py 
once we have the test, I can help look into refactoring this function?

Reply:
Done. LMK what you think.